### DISCUSSION TITLE
Hide() child in lipgloss/tree

### DISCUSSION THREAD

**Discussion by marcotrosi** [charmbracelet/lipgloss#452] [Q&A]
Hello reader,

# Question 1

## TLDR
how can I hide a child/leaf in a tree without using Root()?

## Long question

I have the following code, where I'm trying to hide the second child.

	var T *tree.Tree = tree.Root("root")
	T.Child("first")
	T.Child("second").Hide(true)
	fmt.Println(T)

I understand why it doesn't work, because Child() returns T and not the second child.

I can work my way around by using another Root()

	var T *tree.Tree = tree.Root("root")
	T.Child("first")
	T.Child(tree.Root("second").Hide(true))
	fmt.Println(T)

Okay, but I'm still wondering how to hide that second child without using another Root node. Is it even possible?

I then tried to get the second child and hide it.

	var T *tree.Tree = tree.Root("root")
	T.Child("first")
	T.Child("second")
	T.Children().At(1).Hide(true)
	fmt.Println(T)

Which gives the following error message

`T.Children().At(1).Hide undefined (type tree.Node has no field or method Hide)`

# Question 2

I have these 4 code snippets with corresponding output ...

	var T *tree.Tree = tree.Root("root")
	T.Child("first")
	fmt.Println(T)

<img width="168" alt="image" src="https://github.com/user-attachments/assets/ea196b92-fa8f-4e28-905d-deb982d99def">

	var T *tree.Tree = tree.Root("root")
	T.Child("first")
	T.Child("second")
	fmt.Println(T)

<img width="164" alt="image" src="https://github.com/user-attachments/assets/f3ced3b8-17d2-4d9b-9862-0bb5997476f9">

	var T *tree.Tree = tree.Root("root")
	T.Child("first")
	T.Child(tree.Root("second").Hide(true))
	fmt.Println(T)

<img width="149" alt="image" src="https://github.com/user-attachments/assets/a21b8e3e-7286-4919-b20e-105f67115a82">

	var T *tree.Tree = tree.Root("root")
	T.Child("first")
	T.Child(tree.Root("second").Hide(true))
	T.Child("third")
	fmt.Println(T)

<img width="153" alt="image" src="https://github.com/user-attachments/assets/51b460f7-c56f-4592-92fe-f960e22033f4">

As you can see in snippet number 3 the line is not drawn correctly, it should look like the output of snippet number 1. Is this something I have to take care of myself, or do I have to live with it, or can this safely be considered a bug?

Thank you very much for your support

--marco

**Comment by bashbunni** [comment]
Hey looking into this - I think there are some bugs that you've pointed out here that we will work on. Thanks for sharing this, it'll be super helpful while we work on fixing this

**Reply by bashbunni** [reply]
Update: relating to question 1, here's an update to the lipgloss tree that will resolve that issue for you https://github.com/charmbracelet/lipgloss/pull/454

Will work on getting your second point fixed as well :)

**Reply by marcotrosi** [reply]
Amazing. Thx

Did I get it right, the following code snippet would work with the update?

```
var T *tree.Tree = tree.Root("root")
T.Child("first")
T.Child("second")
T.Children().At(1).Hide(true) // <- this
fmt.Println(T)
```

**Reply by bashbunni** [reply]
With the changes included in the PR, you would use `SetHidden` instead of `Hide`. `Hide` is for when you're setting a tree to be hidden when it's being created or in other cases where you may want to return that tree

**Reply by marcotrosi** [reply]
Sorry for my hopefully not too annoying questions, but it's my first Go project.

My use case would indeed be at creation time.
Like in this example which didn't work and probably still won't after the PR, because Child() returns the tree.

```
var T *tree.Tree = tree.Root("root")
T.Child("first")
T.Child("second").Hide(true)
fmt.Println(T)
```

If I got you right then I would have to create the Child() first, then get the Children() from the tree, get the last child which I just created, and then use SetHidden(). Basically this ...

```
var T *tree.Tree = tree.Root("root")
T.Child("first")
T.Child("second")
T.Children().At(len(T.Children())-1).SetHidden(true)
fmt.Println(T)
```

Am I missing a simpler solution?
Or I use my workaround where I only use Root() calls, then I can use Hide() directly. I anyway don't understand why there is distinction between Root and Child, and not everything is a Node and either it has children or it hasn't.

**Reply by bashbunni** [reply]
Hey @marcotrosi sorry for the delay on this, here's the functionality that has been merged to `master` https://github.com/charmbracelet/lipgloss/blob/master/tree/example_test.go

The examples in that file should answer all of your questions about usage, but if anything is still unclear, please don't hesitate to let us know. PS: your questions have been really helpful in identifying gaps in the tree implementation - definitely not a bother! Thank you for bringing this to our attention 🙏

**Reply by marcotrosi** [reply]
Hello @bashbunni 
thank you so much. I will have a look at the example and test it. Thank you thank you thank you. sending ghost hug.

**Reply by marcotrosi** [reply]
@bashbunni 
I don't understand why the Leaf type exists. Is it needed for different drawing style of Leaf vs. normal TreeNode with empty list of Children? But okay, not so important.
The only thing I find inconsistent is that NewLeaf() is the only function that has a boolean parameter for the Hidden value. Would be super helpful if the other Node creation function(s) would have that too. Is that possible?

**Comment by marcotrosi** [comment]
Hello @bashbunni 

I'm back with some news.
I started using the updated version, tested it, it seemed to work, and then the other day I found another issue with the rendering.
See image here ...
<img width="798" alt="bug" src="https://github.com/user-attachments/assets/dcd6cce7-7a5f-4857-ab94-0142dffa5caa" />
As you can see the line extends to a hidden node.

Then I thought I'll look at the implementation of the tree package, but my eyes are not used to reading Golang code and also I never learned Go, so I decided to write my own tree to see if I can find a way without any drawing issues.

I guess I found a smart way to create the lines in a tree without issues.

Here is an image that showcases what the code can do.
Please ignore the rest around, I only wrote some functions so that I can fill the tree with data, but there are still some things I haven't figured out in Golang, like optional parameters, that's why you see everywhere the true/false for hidden or not hidden nodes.
![overview](https://github.com/user-attachments/assets/9de504a3-9037-4a05-bb15-6c83447021e5)

So you see with just a few lines I can render the trees in different ways. As examples I implemented 4 different styles. For the last folder style one I introduced a hidechildren boolean, which I need for displaying open or closed folders.

The concept is very simple. I have a FILO stack that I pass to the rendering functions, and therefore I know for each line that I want to render exactly which characters to draw, because I can calculate/determine whatever I need with that data.

If you want my concept code please let me know.

With that I wanna say thank you for all your support, and I think we can close now this issue.

All the best

--marco
